### DISCUSSION TITLE
Dynamically set install dir in WiX V4 from environment variable

### DISCUSSION THREAD

**Discussion by domischell** [wixtoolset/issues#7278] [Questions] [answered]
Hi,
I want to install a component under a directory that is defined at installation-time by an environment variable on the machine, for example SOME_ENV_VAR. I found a [question on stackoverflow](https://stackoverflow.com/questions/34993486/dynamically-set-install-dir-in-wix-from-environment-variable) that said to use the following contruct in my Package.wxs:
```xml
<Directory Id="FolderFromEnvVar">
    <Directory Id="InstallMeHere" Name="Subfolder" />
</Directory>
<SetDirectory Id="FolderFromEnvVar" Value="[%SOME_ENV_VAR]" Sequence="first" />
```

I consume this in my XXXComponents.wxs like so
```xml
<Wix>
    <Fragment>
        <Component Id="XXXComponents" Guid="someGUID" Directory="InstallMeHere">
            <File Source=...
```

I open an elevated command shell, set the enviroment variable and run msiexec with my installer
```cmd
set SOME_ENV_VAR="C:\Program Files\Here\Please"
msiexec /i Setup\bin\x64\Debug\Setup.msi /lv %temp%\Installer.txt /quiet
```

And for some reason the files are installed to F:  
The log says
```
MSI (s) (A8:18) [17:56:27:873]: PROPERTY CHANGE: Adding FolderFromEnvVar property. Its value is 'F:\'.
MSI (s) (A8:18) [17:56:27:873]: PROPERTY CHANGE: Adding InstallMeHere property. Its value is 'F:\Subfolder\'.
```

Why is it set to F:\ instead of the value in SOME_ENV_VAR? Any help would be highly appreciated.
Dom

**Comment by robmen** [comment, marked as answer]
My first guess would be that your `set` command is only affecting the local command-prompt and not getting to the server side of the Windows Installer. Try setting the variable in the System on Windows and see if that makes a difference.

If not, dig through the log file a bit more to see what the Windows Installer thinks that variable is. If the env var doesn't show up in the log, try using a `SetProperty` to get a property change log message with the value.

**Reply by domischell** [reply]
Thanks for the hint. The MSI installation indeed doesn't seem to be executed under the current shell and therefore can't see the environment variable that I have set with `set`. It only sees them if specified under the System or User context in the "Advanced System Settings -> Environment Variables...".  I have to use **`setx`** instead of `set`.

```cmd
> setx SOME_USER_VAR "C:\Program Files\Here\Please"

SUCCESS: Specified value was saved.
> msiexec /i Setup\bin\x64\Debug\Setup.msi /lv %temp%\Installer.txt /quiet
```

Now I see the change in the log and the files are installed where I expect them to be
```
MSI (s) (CC:24) [17:30:39:165]: PROPERTY CHANGE: Modifying FolderFromEnvVar property. Its current value is 'F:\'. Its new value: 'C:\Program Files\Here\Please\'.
```

Thanks,
Dom
